packages/protocol/contracts/stability/BreakerBox.sol

```diff
@@ -0,0 +1,305 @@
+   * @return An ordered list of breakers.
+   */
+  function getBreakers() external view returns (address[] memory) {
+    return breakers.getKeys();
+  }
+
+  /**
+   * @notice Checks whether a breaker with the specifed address has been added.
+   */
+  function isBreaker(address breaker) public view returns (bool) {
+    return breakers.contains(breaker);
+  }
+
+  /**
+   * @notice Returns addresses of exchanges that have been added.
+   */
+  function getExchanges() external view returns (address[] memory) {
+    return exchanges;
+  }
+
+  /* ==================== Check Breakers ==================== */
+
+  /**
+   * @notice Checks breakers for a specified exchange to determine the trading mode.
+   * @param exchangeAddress The address of the exchange to run the checks for.
+   * @return currentTradingMode Returns an int representing the current trading mode
+   *                            for the specified exchange.
+   */
+  function checkBreakers(address exchangeAddress) external returns (uint256 currentTradingMode) {
+    TradingModeInfo memory info = exchangeTradingModes[exchangeAddress];
+
+    // Last updated should always have a value gt 0
+    // if it doesn't we can assume this exchange has not been added.
+    if (info.lastUpdatedTime == 0) {
+      return 0;
+    }
+
+    // Check if a breaker has non default trading mode & try to reset
+    if (info.tradingMode != 0) {
+      IBreaker breaker = IBreaker(tradingModeBreaker[info.tradingMode]);
+
+      uint256 cooldown = breaker.getCooldown();
+      bool tryReset = ((cooldown > 0) && (cooldown + info.lastUpdatedTime) <= block.timestamp);
+
+      if (tryReset) {
+        bool canReset = breaker.shouldReset(exchangeAddress);
+        if (canReset) {
+          info.tradingMode = 0;
+          info.lastUpdatedTime = uint64(block.timestamp);
+          info.lastUpdatedBlock = uint128(block.number);
+          exchangeTradingModes[exchangeAddress] = info;
+          emit ResetSuccessful(exchangeAddress, address(breaker));
+          return info.tradingMode;
+        } else {
+          emit ResetAttemptCriteriaFail(exchangeAddress, address(breaker));
+          return info.tradingMode;
+        }
+      } else {
+        emit ResetAttemptNotCool(exchangeAddress, address(breaker));
+        return info.tradingMode;
+      }
+    }
+
+    address[] memory _breakers = breakers.getKeys();
+
+    // Check all breakers
+    for (uint256 i = 0; i < _breakers.length; ++i) {
+      IBreaker breaker = IBreaker(_breakers[i]);
+      bool tripBreaker = breaker.shouldTrigger(exchangeAddress);
+      if (tripBreaker) {
+        info.tradingMode = breakerTradingMode[address(breaker)];
+        info.lastUpdatedTime = uint64(block.timestamp);
+        info.lastUpdatedBlock = uint128(block.number);
+        exchangeTradingModes[exchangeAddress] = info;
+        emit BreakerTripped(address(breaker), exchangeAddress);
+        return info.tradingMode;
```

Comment:
What should happen if multiple breakers should fire at once? Currently we'll just go with the first one that triggers.

Reply:
That is by design. The breakers will be ordered by severity/priority.